.studio/01_base.sh

```diff
@@ -0,0 +1,288 @@
+# code, but also make that documentation available to end user in the studio.
+#
+# The first line of the document will show up when the user runs 'describe'.
+# The full content will appear when the user runs 'describe <function>'
+#
+# When running in a CI environment, these documents are not created.
+document() {
+  name="$1"
+  content=$(cat /dev/stdin)
+
+  [[ -z "${CI:-}" ]] && echo -e "$content" > "/tmp/docs/${name}"
+}
+
+# This is a helper function that allows you to quickly and easily add an alias for
+# a helper function. It will also ensure that a list of aliases are printed when
+# you run `describe`.
+add_alias() {
+  fun="$1"
+  a="$2"
+
+  eval "alias $a=$fun"
+  [[ -z "${CI:-}" ]] && echo -e "$a" > "/tmp/aliases/$fun"
+}
+
+# This function allows the user to learn more about the studio.
+#
+#   1. Passing no argument will print out a list of all documented functions
+#   2. Passing the name of a function will print out that functions documentation
+describe() {
+  if [[ -z "${CI:-}" ]]; then
+    echo
+    if [[ -z "$1" ]]; then
+      echo "The following functions are available for your use in this studio:"
+      echo
+      for file in /tmp/docs/*
+      do
+        fun="$(basename "$file")"
+        echo -n "  $fun"
+        [[ -f "/tmp/aliases/$fun" ]] && printf " [alias: %s]" "$(cat /tmp/aliases/"$fun")"
+        echo
+        echo "  $(head -n 1 "$file")"
+      done
+      if [[ -d "/tmp/aliases" ]]; then
+        echo
+        echo "ALIASES:"
+        for a in /tmp/aliases/*
+        do
+          printf "  %-6s Alias for: %s \\n" "$(cat "$a")" "$(basename "$a")"
+        done
+      fi
+      echo
+      echo "To learn more about a particular function, run 'describe <function>'."
+    else
+      cat "/tmp/docs/$1"
+    fi
+    echo
+  fi
+}
+
+getting_started() {
+  [[ -z "${CI:-}" ]] && cat /dev/stdin
+}
+
+# Use `exit_with` only in very critical situations where you need to
+# exit completely from the studio
+exit_with() {
+  error "$1"
+  exit "$2"
+}
+
+document "install_if_missing" <<DOC
+  Install the package if it is not installed already and binlink the binary to
+  the latest installed package.
+  @(arg:1) The name of the package that contains the binary (e.g., core/musl)
+  @(arg:2) The name of the binary to binlink (e.g., musl-gcc)
+DOC
+install_if_missing() {
+  if [ "$#" -ne 2 ]; then
+    error "Wrong number of arguments to ${FUNCNAME[0]}"
+    describe "${FUNCNAME[0]}"
+    return 1
+  fi
+
+  # Install the package if it is not installed
+  if [[ ! -d "/hab/pkgs/$1" ]]; then
+    hab pkg install "$1" > /dev/null
+  fi
+
+  # Ensure we are binlinking to the same version `hab pkg exec` would run
+  hab pkg binlink --force "$1" "$2" > /dev/null
+}
+
+document "wait_for_success" <<DOC
+  Wait for the given command to succeed.
+  @(arg:*) The command to run
+  Environment Variable:
+    TIMEOUT - The time in seconds to check until failing
+  Example 1
```

Comment:
Would help to have a blank line before this "Example 1" header for readability.
Same for "Example 2" below.